### DISCUSSION TITLE
Hi Rector team! Alex here, would you ever consider adding rector rules for migrating Joomla extensions from 3.10 to 4.1.x?

### DISCUSSION THREAD

**Discussion by alexandreelise** [rectorphp/rector#7166] [Q&A]
Hi, it's quite a big topic in the Joomla! community I was wondering if you could add it to your todo list.
Let me clarify. The predicate is that the migrated extension should abide to Joomla! conventions in both sides 3.10 and 4.1.x otherwise it would be too problematic to find a migration path.
Joomla 3.x conventions: https://docs.joomla.org/J3.x:File_Structure_and_Naming_Conventions
Joomla 4.x conventions: https://docs.joomla.org/J4.x:File_Structure_and_Naming_Conventions

Joomla 4.x BC breaks: https://docs.joomla.org/Potential_backward_compatibility_issues_in_Joomla_4

Can you please shed a light on this one? Have a great day and nice weekend. @TomasVotruba and the rest of the team.

**Comment by TomasVotruba** [comment]
Hi Alex,

thanks for your sources. 
You can implement those rules yourself. The best way to start is documentation: https://github.com/rectorphp/rector/tree/main/docs
If you plan to do such upgrade in the long-term, I can definitelly recommend the book to get deeper knowledge and answers to questions: https://getrector.org/book

We can offer our services to make it happen faster. We already helps few open-source and private projects with ugprades set. In rewards, they sponsor us back on Github Sponsors or privately supports Rector. These funds are important, as Rector is fully open-source including all its features and this is the main source of funding it back.

**Reply by alexandreelise** [reply]
Hi Tomas, 
Thanks for your quick reply. I already bought your book. One question remains is it because Joomla is not mainstream like the other frameworks rector officially supports? @TomasVotruba

**Reply by alexandreelise** [reply]
Thanks anyway. I'll do my best to create custom rector rules digging in the heart and soul of rector.

**Reply by TomasVotruba** [reply]
Basically Rector was born on Symfony and Nette components, so it supported them first. It also uses PHP :) and PHPUnit, so does most of the MVC PHP community, so that was initial support we kicked out. The rest of package we support now is based on activity of PHP community and private support.

Good luck with the upgrade :+1:

**Reply by alexandreelise** [reply]
Thank Tomas. Now I understand better. Will read the book more deeply and try to understand how everything works. Thanks for that awesome tool. BTW I saw some of your talks on Rector on Youtube and heard some of your podcast intervention on rector when you explain that it came from being "lazy" It's an awesome tool that deserves it's place in the PHP Community. Thanks again to you and your team. Take care.

**Reply by TomasVotruba** [reply]
Thank you for such specific words of appreciation :bow:  I'm honored, thank you
